tools/cabana/streams/routes.cc

```diff
@@ -79,45 +96,78 @@ void RoutesDialog::fetchRoutes() {
   if (device_list_->currentIndex() == -1 || device_list_->currentData().isNull())
     return;
 
-  route_list_->clear();
-  route_list_->setEmptyText(tr("Loading..."));
+  currentRoutesList()->clear();
+  currentRoutesList()->setEmptyText(tr("Loading..."));
 
   HttpRequest *http = new HttpRequest(this, !Hardware::PC());
   QObject::connect(http, &HttpRequest::requestDone, this, &RoutesDialog::parseRouteList);
 
   // Construct URL with selected device and date range
   auto dongle_id = device_list_->currentData().toString();
-  QDateTime current = QDateTime::currentDateTime();
-  QString url = QString("%1/v1/devices/%2/routes_segments?start=%3&end=%4")
-                    .arg(CommaApi::BASE_URL).arg(dongle_id)
-                    .arg(current.addDays(-(period_selector_->currentData().toInt())).toMSecsSinceEpoch())
-                    .arg(current.toMSecsSinceEpoch());
-  http->sendRequest(url);
+  if(isPreservedTabSelected()) {
+    QString url = QString("%1/v1/devices/%2/routes/preserved")
+                      .arg(CommaApi::BASE_URL)
+                      .arg(dongle_id);
+    http->sendRequest(url);
+  } else {
+    QDateTime current = QDateTime::currentDateTime();
+    QString url = QString("%1/v1/devices/%2/routes_segments?start=%3&end=%4")
+                      .arg(CommaApi::BASE_URL)
+                      .arg(dongle_id)
+                      .arg(current.addDays(-(period_selector_->currentData().toInt())).toMSecsSinceEpoch())
+                      .arg(current.toMSecsSinceEpoch());
+    http->sendRequest(url);
+  }
 }
 
 void RoutesDialog::parseRouteList(const QString &json, bool success, QNetworkReply::NetworkError err) {
   if (success) {
     for (const QJsonValue &route : QJsonDocument::fromJson(json.toUtf8()).array()) {
-      uint64_t start_time = route["start_time_utc_millis"].toDouble();
-      uint64_t end_time = route["end_time_utc_millis"].toDouble();
-      auto datetime = QDateTime::fromMSecsSinceEpoch(start_time);
-      auto item = new QListWidgetItem(QString("%1    %2min").arg(datetime.toString()).arg((end_time - start_time) / (1000 * 60)));
+      uint64_t to;
+      QDateTime from;
+
+      if(isPreservedTabSelected()) {
+        QString start_time = route["start_time"].toString();
+        QString end_time = route["end_time"].toString();
+
+        from = QDateTime::fromString(start_time, Qt::ISODateWithMs);
+        to = QDateTime::fromString(end_time, Qt::ISODateWithMs).msecsTo(from);
+      } else {
+        uint64_t start_time = route["start_time_utc_millis"].toDouble();
+        uint64_t end_time = route["end_time_utc_millis"].toDouble();
+        from = QDateTime::fromMSecsSinceEpoch(start_time);
+        to = (end_time - start_time);
+      }
+
+      QListWidgetItem *item = new QListWidgetItem(QString("%1    %2min").arg(from.toString()).arg(to / (1000 * 60)));
```

Comment:
how about:
``` C++
QDateTime from, to;
if (isPreservedTabSelected()) {
  from = QDateTime::fromMSecsSinceEpoch(route["start_time_utc_millis"].toDouble());
  to = QDateTime::fromMSecsSinceEpoch(route["end_time_utc_millis"].toDouble());
} else {
  from = QDateTime::fromString(route["start_time"].toString(), Qt::ISODateWithMs);
  to = QDateTime::fromString(route["end_time"].toString(), Qt::ISODateWithMs);
}
auto item = new QListWidgetItem(QString("%1    %2min").arg(from.toString()).arg(from.secsTo(to) / 60));
...
}
```
it's more concise, readable, eliminates unnecessary variable assignments.